introduce MixedMesh class

ksagiyam opened the pull request:
Introduce `MixedMesh` class for multi-mesh problems.

Note that `extract_domains()` now uses more robust `sort_domains()` inside, so it might behave slightly differently.

ksagiyam commented on ufl/domain.py:
```diff
@@ -126,6 +145,69 @@ def _ufl_sort_key_(self):
+
+@attach_ufl_id
+class MixedMesh(AbstractDomain, UFLObject):
+    """Symbolic representation of a mixed mesh."""
```
doctoring

ksagiyam commented:
@jpdean @jorgensd This is what we briefly discussed at PDESoft. Could you have a look?

jorgensd commented on ufl/algorithms/analysis.py:
```diff
@@ -198,19 +200,22 @@ def extract_base_form_operators(a):
-    base_coeff_and_args = extract_type(a, (BaseArgument, BaseCoefficient))
-    arguments = [f for f in base_coeff_and_args if isinstance(f, BaseArgument)]
-    coefficients = [f for f in base_coeff_and_args if isinstance(f, BaseCoefficient)]
+    base_coeff_and_args_and_gq = extract_type(a, (BaseArgument, BaseCoefficient, GeometricQuantity))
```
Could we just call this list `base_types`, as it is quite explicit from the function call what it does, and it makes the latter code more readable.

ksagiyam replied:
Renamed it `terminals`.

jorgensd commented on ufl/algorithms/analysis.py:
```diff
@@ -226,20 +231,36 @@ def extract_arguments_and_coefficients(a):
+    # Build count: instance mappings, should be one to one
+    gqcounts = {}
+    for gq in geometric_quantities:
+        if not isinstance(gq._domain, Mesh):
+            raise TypeError(f"{gq}._domain must be a Mesh: got {gq._domain}")
```
What else com `gq._domain` be? is this relying on the deprecated definition that a `domain` can be defined just through a `ufl.Cell`?

ksagiyam replied:
No. It checks if `gq._domain` is `Mesh` and not `MixedMesh`.

jorgensd commented on test/test_mixed_function_space_with_mixed_mesh.py:
```diff
@@ -0,0 +1,81 @@
+    mesh0 = Mesh(FiniteElement("Lagrange", cell, 1, (2,), identity_pullback, H1), ufl_id=100)
+    mesh1 = Mesh(FiniteElement("Lagrange", cell, 1, (2,), identity_pullback, H1), ufl_id=101)
+    mesh2 = Mesh(FiniteElement("Lagrange", cell, 1, (2,), identity_pullback, H1), ufl_id=102)
+    domain = MixedMesh(mesh0, mesh1, mesh2)
+    V = FunctionSpace(domain, elem)
```
So we now have two ways of creating a `FunctionSpace` with `MixedElement`s.
1. A single mesh of class `Mesh` with a `MixedElement`, where `MixedElement` can have arbitrary many elements
2. A `MixedMesh` consisting of `M` meshes, and a `MixedElement` consisting of `M` elements.

Do we need additional checks in the `FunctionSpace` constructor to ensure that this is satisfied (instead of catching this at a later instance, inside for instance `GenericDerivativeRuleset.reference_value`.

I would also like to highlight that the API for `MixedElement` and `MixedMeshes` differs, as one passes a list, while the other unrolls a list, I think we should be consistent, and use lists in both places

ksagiyam replied:
Added checks in `FunctionSpace` constructor and changed `MixedMesh` API as suggested.  Thanks.

ksagiyam commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -655,16 +668,52 @@ def reference_value(self, o):
+            if element.num_sub_elements != len(domain):
+                raise RuntimeError(f"{element.num_sub_elements} != {len(domain)}")
+            g = ReferenceGrad(o)
+            vsh = g.ufl_shape[:-1]
```
Turned out `vsh` could be removed, so I did.

You are right. Currently, `MixedMesh` can only be composed of meshes of the same cell_type, so topological dimension is well defined. I edited the doc string of `MixedMesh` to make this point clear.

ksagiyam commented:
Updated PR summary.

Added more checks.

Fixed `MixedPullback.physical_value_shape()`.

ksagiyam commented:
Can I have another round of review on this?

jorgensd commented on ufl/algorithms/analysis.py:
```diff
@@ -198,19 +200,20 @@ def extract_base_form_operators(a):
 
     Args:
         a: A BaseForm, Integral or Expr
     """
```
Could we add typing of input/output here?

ksagiyam replied:
Done.

jorgensd commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -84,6 +86,17 @@
 # - ReferenceDivRuleset
 
 
+def flatten_domain_element(domain, element):
+    """Return the flattened (domain, element) pairs for mixed domain problems."""
+    if not isinstance(domain, MixedMesh):
+        return ((domain, element),)
+    flattened = ()
+    assert len(domain) == len(element.sub_elements)
+    for d, e in zip(domain, element.sub_elements):
+        flattened += flatten_domain_element(d, e)
+    return flattened
```
Typing + specification of what is returned if the problem is not a "mixed domain problem"/

jorgensd commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -657,16 +670,51 @@ def reference_value(self, o):
+            ref_dim = g.ufl_shape[-1]
+            components = []
+            dofoffset = 0
+            for d, e in flatten_domain_element(domain, element):
+                esh = e.reference_value_shape
+                if esh == ():
+                    esh = (1,)
+                if isinstance(e.pullback, PhysicalPullback):
+                    if ref_dim != self._var_shape[0]:
+                        raise NotImplementedError("""
+                            PhysicalPullback not handled for immersed domain :
+                            reference dim ({ref_dim}) != physical dim (self._var_shape[0])""")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(ref_dim):
+                            components.append(g[(dofoffset + idx,) + (i,)])
+                else:
+                    K = JacobianInverse(d)
+                    rdim, gdim = K.ufl_shape
+                    if rdim != ref_dim:
+                        raise RuntimeError(f"{rdim} != {ref_dim}")
+                    if gdim != self._var_shape[0]:
+                        raise RuntimeError(f"{gdim} != {self._var_shape[0]}")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(gdim):
+                            temp = Zero()
+                            for j in range(rdim):
+                                temp += g[(dofoffset + idx,) + (j,)] * K[j, i]
+                            components.append(temp)
+                dofoffset += int(np.prod(esh))
+            return as_tensor(np.asarray(components).reshape(g.ufl_shape[:-1] + self._var_shape))
```
Does one need the conditional when `flatten_domain_element` works for non-mixed meshes.

ksagiyam replied:
Shape of `g` (which I now call `rgrad`) is slightly different for scalar and non-scalar spaces, e.g. `(3,)` v.s. `(5, 3)` for `topological dim = 3`, so `g[(dofoffset + idx,) + (j,)]` fails for scalar spaces. We could special case for the scalar space inside the `flatten_domain_element` loop, but I thought it would be cleaner if we just did that outside.

jorgensd commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -657,16 +670,51 @@ def reference_value(self, o):
+            ref_dim = g.ufl_shape[-1]
+            components = []
+            dofoffset = 0
+            for d, e in flatten_domain_element(domain, element):
+                esh = e.reference_value_shape
+                if esh == ():
+                    esh = (1,)
+                if isinstance(e.pullback, PhysicalPullback):
+                    if ref_dim != self._var_shape[0]:
+                        raise NotImplementedError("""
+                            PhysicalPullback not handled for immersed domain :
+                            reference dim ({ref_dim}) != physical dim (self._var_shape[0])""")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(ref_dim):
+                            components.append(g[(dofoffset + idx,) + (i,)])
+                else:
+                    K = JacobianInverse(d)
+                    rdim, gdim = K.ufl_shape
+                    if rdim != ref_dim:
+                        raise RuntimeError(f"{rdim} != {ref_dim}")
+                    if gdim != self._var_shape[0]:
+                        raise RuntimeError(f"{gdim} != {self._var_shape[0]}")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(gdim):
+                            temp = Zero()
+                            for j in range(rdim):
+                                temp += g[(dofoffset + idx,) + (j,)] * K[j, i]
+                            components.append(temp)
+                dofoffset += int(np.prod(esh))
```
Some documentation of the logic within this loop would be nice.
I guess the idea is:
- For a mixed domain problem, accumulate all derivatives in expanded index notation.
Please correct me if I am mistaken.

I think one thing that would help is to define `np.prod(esh)` as a variable, as it is used in multiple loops and to accumulate the counter.

Also, I think you do not need the
```
if esh == ():
    esh = (1,)
```
as `int(np.prod(()))` ref: https://numpy.org/doc/2.1/reference/generated/numpy.prod.html
> The product of an empty array is the neutral element 1

One could add a guard against future changing behavior by adding
```python
esh = e.reference_value_shape
num_components = int(np.prod(esh))
assert num_components > 0
```

ksagiyam replied:
You are right. Simplified as suggested and added more comments. Thanks.

jorgensd commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -657,16 +670,51 @@ def reference_value(self, o):
+                                temp += g[(dofoffset + idx,) + (j,)] * K[j, i]
+                            components.append(temp)
+                dofoffset += int(np.prod(esh))
+            return as_tensor(np.asarray(components).reshape(g.ufl_shape[:-1] + self._var_shape))
```
Is using `numpy.asarray` safe here?

ksagiyam replied:
I think so.

jorgensd commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -678,10 +726,43 @@ def reference_grad(self, o):
+            element = f.ufl_function_space().ufl_element()
+            if element.num_sub_elements != len(domain):
+                raise RuntimeError(f"{element.num_sub_elements} != {len(domain)}")
+            g = ReferenceGrad(o)
```
Same comment as for previous function

ksagiyam replied:
Made similar changes here, too.

jorgensd commented on ufl/algorithms/compute_form_data.py:
```diff
@@ -262,6 +262,14 @@ def compute_form_data(
+            all_domains = extract_domains(integral.integrand(), expand_mixed_mesh=False)
+            if any(isinstance(m, MixedMesh) for m in all_domains):
+                raise NotImplementedError("""
+                    Only integral_type="cell" can be used with MixedMesh""")
```
Maybe add what cell type you got as input to error message.

ksagiyam replied:
Elaborated the error message.

jorgensd commented on ufl/domain.py:
```diff
@@ -50,6 +50,33 @@ def topological_dimension(self):
+
+    def can_make_function_space(self, element):
+        """Check whether this mesh can make a function space with ``element``."""
+        raise NotImplementedError("check_compatibility() method not implemented")
```
```suggestion
        raise NotImplementedError("can_make_function_space() method not implemented")
```

jorgensd commented on ufl/domain.py:
```diff
@@ -126,6 +153,119 @@ def _ufl_sort_key_(self):
+        """Return the component meshes."""
+        return (self,)
+
+    def iterable_like(self, element):
```
Typing-hinting for new functions in UFL would be nice.

jorgensd replied:
This should be type-hinted as it has assumptions on the fact that element has the property `sub_elements`.

jorgensd replied:
Why not use `num_sub_elments` as in `mixedmesh`?

ksagiyam replied:
Added type-hinting. Right, we now use `num_sub_elements`.

jorgensd commented on ufl/domain.py:
```diff
@@ -126,6 +153,119 @@ def _ufl_sort_key_(self):
+        return self._meshes
+
+    def iterable_like(self, element):
+        """Return iterable object that is iterable like ``element``."""
```
Assumes that you have `num_sub_elements`, so should be type-hinted.

ksagiyam replied:
Fixed. `AbstractFiniteElement` declares `num_sub_elements`.

jorgensd commented on ufl/domain.py:
```diff
@@ -126,6 +153,119 @@ def _ufl_sort_key_(self):
+                element.num_sub_elements ({element.num_sub_elements})""")
+        return self
+
+    def can_make_function_space(self, element):
```
Type-hinting.

ksagiyam replied:
Done.

jorgensd commented on ufl/domain.py:
```diff
@@ -190,26 +330,34 @@ def as_domain(domain):
 
 
-def join_domains(domains):
+def join_domains(domains, expand_mixed_mesh=True):
```
Add comment about what expand mixed mesh true/false means + type-hinting.

ksagiyam replied:
Done.

jorgensd commented on ufl/domain.py:
```diff
@@ -226,20 +374,26 @@ def join_domains(domains):
 
 
-def extract_domains(expr):
+def extract_domains(expr, expand_mixed_mesh=True):
```
Add documentation regarding keyword argument `expand_mixed_mesh`.

jorgensd commented on ufl/domain.py:
```diff
@@ -226,20 +374,26 @@ def join_domains(domains):
+def extract_unique_domain(expr, expand_mixed_mesh=True):
     """Return the single unique domain expression is defined on or throw an error."""
-    domains = extract_domains(expr)
+    domains = extract_domains(expr, expand_mixed_mesh=expand_mixed_mesh)
```
Add kwarg docstring

ksagiyam replied:
Done.

jhale commented:
Copied from Slack:

The implementation looks OK to me (quick look), but the tests look a bit light. In terms of the naming, I was also confused by the use of Mixed. At least when we’re talking about MixedElement, aside from it having historical meaning, my impression was that we are talking about using the Cartesian product on the elements of the list and maintaining ordering. My understanding of MixedMesh is that we are maintaining the same ordering, but I didn’t get the impression it was a Cartesian product - consequently, the used of Mixed wasn’t clear to me. This is aside from the idea that Mixed Mesh has no traditional meaning in the literature - when the idea was first proposed I thought it was related to having a mesh with mixed cell type, to illustrate the confusion.

Of course these are ‘misconceptions’ on my part, but I’m just trying to get across the confusion that appeared in my own mind, so we can discuss if different naming of the same concept might be appropriate.

connorjward reviewed (commented):
I am happy to be argued down but personally I think the abstraction is a bit strange. Why can we not have, instead of
```py
    cell = triangle
    elem0 = FiniteElement("Lagrange", cell, ...)
    elem1 = FiniteElement("Brezzi-Douglas-Marini", cell, ...)
    elem2 = FiniteElement("Discontinuous Lagrange", cell, ...)
    elem = MixedElement([elem0, elem1, elem2])
    mesh0 = Mesh(FiniteElement("Lagrange", cell, ...))
    mesh1 = Mesh(FiniteElement("Lagrange", cell, ...))
    mesh2 = Mesh(FiniteElement("Lagrange", cell, ...))
    domain = MixedMesh([mesh0, mesh1, mesh2])
    V = FunctionSpace(domain, elem)
```
something like
```py
    cell = triangle
    elem0 = FiniteElement("Lagrange", cell, ...)
    mesh0 = Mesh(FiniteElement("Lagrange", cell, ...))
    V0 = FunctionSpace(mesh0, elem0)

    elem1 = FiniteElement("Brezzi-Douglas-Marini", cell, ...)
    mesh1 = Mesh(FiniteElement("Lagrange", cell, ...))
    V1 = FunctionSpace(mesh1, elem1)

    elem2 = FiniteElement("Discontinuous Lagrange", cell, ...)
    mesh2 = Mesh(FiniteElement("Lagrange", cell, ...))
    V2 = FunctionSpace(mesh2, elem2)

    V = MixedFunctionSpace([V0, V1, V2])
```
where we have a `MixedFunctionSpace` instead of a `MixedMesh`.

More generally what is the difference between a `FunctionSpace` with a `MixedElement` and a `MixedFunctionSpace` using non-`Mixed` elements?

jorgensd commented:
> ```python
>     cell = triangle
>     elem0 = FiniteElement("Lagrange", cell, ...)
>     mesh0 = Mesh(FiniteElement("Lagrange", cell, ...))
>     V0 = FunctionSpace(mesh0, elem0)
> 
>     elem1 = FiniteElement("Brezzi-Douglas-Marini", cell, ...)
>     mesh1 = Mesh(FiniteElement("Lagrange", cell, ...))
>     V1 = FunctionSpace(mesh1, elem1)
> 
>     elem2 = FiniteElement("Discontinuous Lagrange", cell, ...)
>   
> ```

This is what we use in DOLFINx at the moment.

I think the current  mixed-mesh abstraction is going to be confusing when we introduce meshes with multiple cell types (ie mix of quadrilateral and triangle, or hex, yet, prism).

ksagiyam commented:
@jorgensd I do not think merely allowing for `Coefficient(MixedFunctionSpace)` will solve all issues; as long as `MixedFunctionSpace` does not subclass `FunctionSpace` and does not have `ufl_domain`, I think we will need to special case functions/methods in UFL for `MixedFunctionSpace` just as `Coefficients` function currently does. As I mentioned, introduction of `MixedMesh` allows UFL to be unaware of the mixed spaces in the most part, and it only needs to special case methods at the very bottom, i.e. `GradRuleset.reference_value()`, `GradRuleset.reference_grad()`, and `pullback.apply()`.

I did not mean nested mixed spaces by "multiple meshes + multiple spaces". I just meant (meshA, elemA) x (meshB, elemB). Though we do not use nested mixed spaces in Firedrake, some functions such as `flatten_domain_element` are written with this in mind as you might have noticed. I do not think consideration of such nested mixed spaces would affect the mixed-problem design in UFL, though.

@jhale I think the use of `Mixed` in `MixedMesh` is consistent with the use of `Mixed` in `MixedElement`. If we consider assembling (i, j)-block of a matrix, we can view it as mixing `mixed_mesh[i]` and `mixed_mesh[j]` as well as mixing `mixed_element[i]` and `mixed_element[j]`. 

But I agree that it is confusing. As @dham suggested, `MeshSequence` or `MultiMesh` is also fine by me.

jorgensd commented:
> @jorgensd I do not think merely allowing for `Coefficient(MixedFunctionSpace)` will solve all issues; as long as `MixedFunctionSpace` does not subclass `FunctionSpace` and does not have `ufl_domain`, I think we will need to special case functions/methods in UFL for `MixedFunctionSpace` just as `Coefficients` function currently does. As I mentioned, introduction of `MixedMesh` allows UFL to be unaware of the mixed spaces in the most part, and it only needs to special case methods at the very bottom, i.e. `GradRuleset.reference_value()`, `GradRuleset.reference_grad()`, and `pullback.apply()`.
> 
> I did not mean nested mixed spaces by "multiple meshes + multiple spaces". I just meant (meshA, elemA) x (meshB, elemB). Though we do not use nested mixed spaces in Firedrake, some functions such as `flatten_domain_element` are written with this in mind as you might have noticed. I do not think consideration of such nested mixed spaces would affect the mixed-problem design in UFL, though.
> 
> @jhale I think the use of `Mixed` in `MixedMesh` is consistent with the use of `Mixed` in `MixedElement`. If we consider assembling (i, j)-block of a matrix, we can view it as mixing `mixed_mesh[i]` and `mixed_mesh[j]` as well as mixing `mixed_element[i]` and `mixed_element[j]`.
> 
> But I agree that it is confusing. As @dham suggested, `MeshSequence` or `MultiMesh` is also fine by me.

Let's not use `MultiMesh`, as that has been used in FEniCS before to mean hierarchies of meshes.

jhale commented:
`MeshSequence` is clearer and at least to my understanding it is consistent with Python's abstract base class terminology `collections.abc.Sequence`.

@ksagiyam  I see your analogy between `MixedMesh` and `MixedElement` at the form level, however, `MixedElement` was named after a widely established and accepted terminology from the finite element literature. We don't have this for `MixedMesh`, and so my preference is to take precise terminology from elsewhere.

ksagiyam commented on ufl/domain.py:
```diff
@@ -190,26 +330,34 @@ def as_domain(domain):
     """Convert any valid object to an AbstractDomain type."""
     if isinstance(domain, AbstractDomain):
         # Modern UFL files and dolfin behaviour
+        (domain,) = set(domain.meshes)
```
I think `as_domain` is more widely used. In Firedrake I can see `as_domain(V)` for instance.

ksagiyam commented on ufl/domain.py:
```diff
@@ -126,6 +153,119 @@ def _ufl_sort_key_(self):
+
+    """
+
+    def __init__(self, meshes, ufl_id=None, cargo=None):
```
Done.

ksagiyam commented:
Thanks, all. Yes, let us use `MeshSequence`. The followings are the changes I have made according to the suggestions:

- Addressed @jorgensd 's suggestions,
- Renamed `MixedMesh` `MeshSequence`,
- Removed `ufl_id` from `MeshSequence`; I realised that my Firedrake branch broke after a recent rebase, and found that it was caused by distinguishing two `MeshSequence` instances made from the same set of `Mesh`es. Thinking about this, I thought `ufl_id` in `MeshSequence` is redundant; please see updated `MeshSequence.__repr__()` and `MeshSequence._ufl_hash_data_()`. I also added a test to highlight this point.

@jhale What sort of tests do you think are missing? After this PR, I will soon make another PR that adds logics to handle restrictions (to handle cases where exterior facet integration of meshA is interior facet integration of meshB, etc). At that stage I will add one or two more examples that test facet integrations.

ksagiyam commented:
@jorgensd Do you have any further comments on this PR by any chance? I would like to have this PR go.

jorgensd commented:
> @jorgensd Do you have any further comments on this PR by any chance? I would like to have this PR go.

I’d like @garth-wells, @chrisrichardson and @michalhabera  to have a look over this. If they don’t have time, I guess we should get towards merging this.

mscroggs commented on ufl/domain.py:
```diff
@@ -50,6 +58,33 @@ def topological_dimension(self):
+        """Return iterable component meshes."""
+        return iter(self.meshes)
+
+    def iterable_like(self, element: AbstractFiniteElement) -> Iterable["Mesh"] | "MeshSequence":
```
Do we need to increase the minimum python version in pyproject.toml to  allow using `|` in typing?
Given the minimum version is currently 3.8 which has passed end-of-life, we should probably bump it anyway, but not necessarily in this PR unless it's needed here

ksagiyam replied:
Thanks. It's probably not a good idea to bump python version in this PR. I just switched to `typing.Union`.
